app/services/flow/flow_state_machine.rb

```diff
@@ -14,24 +14,82 @@ def index
     end
 
     def show
-      step = current_step
-      analytics.track_event(analytics_visited, step: step) if @analytics_id
-      Funnel::DocAuth::RegisterStep.new(user_id, issuer).call(step, :view, true)
-      register_campaign
-      render_step(step, flow.flow_session)
+      flow_handler = flow.handler(current_step).new(flow)
+
+      if flow_handler.async?
+        # binding.pry
+        async_show(flow_handler)
+      else
+        begin_step
+      end
     end
 
     def update
-      step = current_step
-      result = flow.handle(step)
-      analytics.track_event(analytics_submitted, result.to_h.merge(step: step)) if @analytics_id
-      register_update_step(step, result)
-      flow_finish and return unless next_step
-      render_update(step, result)
+      flow_handler = flow.handler(current_step).new(flow)
+
+      if flow_handler.async?
+        async_update(flow_handler)
+      else
+        result = flow.handle(current_step)
+        end_step(result)
+      end
     end
 
     private
 
+    def async_show(flow_handler)
+      async_state = flow_handler.async_state
+      case async_state.status
+      when :none
+        begin_step
+      when :in_progress
+        redirect_to send(@step_url, step: current_step)
+      when :timed_out
+        begin_step
+      when :done
+        result = flow_handler.after_call(async_state.pii, async_state.result)
+        flow_handler.mark_step_complete(current_step) if result.success?
+        flow_handler.delete_async unless result.success?
+        end_step(result)
+      end
+    end
+
+    def async_update(flow_handler)
+      case flow_handler.async_state.status
+      when :none
+        flow.handle(current_step)
+        # binding.pry
+        flow_handler.mark_step_incomplete(current_step)
+        redirect_to send(@step_url, step: current_step) and return
```

Comment:
all these `and return` seem redundant to me?

Reply:
they probably are, I'll remove them